src/python/pants/engine/environment.py

```diff
@@ -16,6 +16,7 @@
 
 name_value_re = re.compile(r"([A-Za-z_]\w*)=(.*)")
 shorthand_re = re.compile(r"([A-Za-z_]\w*)")
+interpolation_re = re.compile(r"^\$\{([A-Za-z_]{1,}[A-Za-z0-9]{0,})\}$")
```

Comment:
I find it confusing that `${VAR}` is expanded but `$VAR` is not. Also, I would expect that `"embedded $VAR here"` to expand the `VAR` in the middle of the string.

With that said, I'm not too sure that this is what we want here, or need. The `Pants` way to do it, is to add the required values as options on the tool used.

Say you have a twine tool. Along with that is options for specifying which version to use (if not the default) and possibly overriding the download url, if needed.
There'll also be the right place to add twine specific configuration, such as username, password, and config files for twine, to setup repository configurations. Those options may pick values for specific options from ENV vars as needed.

Reply:
Ps.
I realize this comes from my own suggestion earlier over in your own repo. Didn't think of the tooling options then, apologies :)
Ds.

Reply:
> I find it confusing that ${VAR} is expanded but $VAR is not. Also, I would expect that "embedded $VAR here" to expand the VAR in the middle of the string.

This is mostly just a result of my idioms, as I'm used to predominantly using the `${foo}` syntax in `Dockerfile` and `docker-compose.yml` files.

Also has the convenient side-effect of supporting non-variable values that start with a `$` symbol. Still doesn't cover `${Not_A_Variable,This_Is_a_password}` but I think the odds of that are significantly smaller.

> With that said, I'm not too sure that this is what we want here, or need. The Pants way to do it, is to add the required values as options on the tool used.
> Say you have a twine tool. Along with that is options for specifying which version to use (if not the default) and possibly overriding the download url, if needed.
> There'll also be the right place to add twine specific configuration, such as username, password, and config files for twine, to setup repository configurations. Those options may pick values for specific options from ENV vars as needed.

The username and password isn't a configuration value for Twine, but rather a configuration value for the repository. In my use-case, I have two destinations for python packages. One for internal tooling, and the other to a "public" repository for clients. Both have different credentials.

Ideally, I wouldn't like to implement it specifically and have it handled by something that resolves #12797. But my familiarity with Pants only goes so far :wink:

Reply:
Usually, there's an escape for `$` itself for disambiguation. That makes it clear what is, and what is not, a variable for expansion.

However, regarding twine, you can provide a configuration file with repositories, and those in turn supports environment variables: https://twine.readthedocs.io/en/latest/#environment-variables

Question then just becomes, what environment variables to pass on when running twine, but I think that will be a cleaner approach.

Reply:
Perhaps it is an OK limitation to not expect publishing to more than a single pypi repo at a time? 
In which case the env variables to use are a defined set at all times.